# Hacker News Project

In this project we will work on a data set of submissions to Hacker News website. We will specifically look into titles that begin with either Ask HN or Show HN. We will try to determine which type of posts receive the more comments on average and also if posts created at a certain time receive more comments on average. 

In [9]:
#first, let's open the dataset and display the header and first five rows without the header

from csv import reader
opened_file=open('hacker_news.csv')
readed_file=reader(opened_file)
hn=list(readed_file)
headers=hn[0]
hn=hn[1:]
print(headers)
print('\n')
print(hn[0:4])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Now, we will start filtering our data. Hence, we will separate the posts beginning with Ask HN and Show HN into two different lists

In [24]:
ask_posts=[]
show_posts=[]
other_posts=[]

for row in hn: 
    title=row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


Next, let's determine if ask posts or show posts receive more comments on average. 

In [26]:
total_ask_comments=0
for row in ask_posts:
    num_comments=int(row[4])
    total_ask_comments+=num_comments
    
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

total_show_comments=0
for row in show_posts:
    num_comments=int(row[4])
    total_show_comments+=num_comments
    
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

14.038417431192661
10.31669535283993


According to the result above, we can say that ask posts receive more comments on average than show posts (14,04 versus 10.32)

In [53]:
import datetime as dt
result_list=[]
counts_by_hour={}
comments_by_hour={}

for row in ask_posts:
    created_at=row[6]
    n_comments=int(row[4])
    my_list=[created_at,n_comments]
    result_list.append(my_list)

for row in result_list:
    hour=row[0]
    hour=dt.datetime.strptime(hour,"%m/%d/%Y %H:%M")
    hour=hour.strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=row[1]
    else:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=row[1]

print(counts_by_hour)
print('\n')
print(comments_by_hour)
    

{'04': 47, '17': 100, '13': 85, '14': 107, '06': 44, '16': 108, '22': 71, '10': 59, '00': 55, '23': 68, '11': 58, '08': 48, '21': 109, '18': 109, '01': 60, '03': 54, '09': 45, '19': 110, '02': 58, '12': 73, '07': 34, '15': 116, '20': 80, '05': 46}


{'04': 337, '17': 1146, '13': 1253, '14': 1416, '06': 397, '16': 1814, '22': 479, '10': 793, '00': 447, '23': 543, '11': 641, '08': 492, '21': 1745, '18': 1439, '01': 683, '03': 421, '09': 251, '19': 1188, '02': 1381, '12': 687, '07': 267, '15': 4477, '20': 1722, '05': 464}


Here above we created two dictionaries which contain the number of asks posts created during each hour of the day and the corresponding number of comments created at each hour of the day.

Now, let's use the two dictionaries to calculate the average number of comments for posts created during each hour of the day. 

In [54]:
avg_by_hour=[]
for hour in counts_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])

print(avg_by_hour)

[['04', 7.170212765957447], ['17', 11.46], ['13', 14.741176470588234], ['14', 13.233644859813085], ['06', 9.022727272727273], ['16', 16.796296296296298], ['22', 6.746478873239437], ['10', 13.440677966101696], ['00', 8.127272727272727], ['23', 7.985294117647059], ['11', 11.051724137931034], ['08', 10.25], ['21', 16.009174311926607], ['18', 13.20183486238532], ['01', 11.383333333333333], ['03', 7.796296296296297], ['09', 5.5777777777777775], ['19', 10.8], ['02', 23.810344827586206], ['12', 9.41095890410959], ['07', 7.852941176470588], ['15', 38.5948275862069], ['20', 21.525], ['05', 10.08695652173913]]


The result above is hard to read. Therefore, let's sort the list of list and print the five highest values in a more convenient format.

In [60]:
swap_avg_by_hour=[]
for row in avg_by_hour:
    first_element=row[0]
    second_element=row[1]
    swap_avg_by_hour.append([second_element, first_element])

print(swap_avg_by_hour)

sorted_swap=sorted(swap_avg_by_hour, reverse=True)

[[7.170212765957447, '04'], [11.46, '17'], [14.741176470588234, '13'], [13.233644859813085, '14'], [9.022727272727273, '06'], [16.796296296296298, '16'], [6.746478873239437, '22'], [13.440677966101696, '10'], [8.127272727272727, '00'], [7.985294117647059, '23'], [11.051724137931034, '11'], [10.25, '08'], [16.009174311926607, '21'], [13.20183486238532, '18'], [11.383333333333333, '01'], [7.796296296296297, '03'], [5.5777777777777775, '09'], [10.8, '19'], [23.810344827586206, '02'], [9.41095890410959, '12'], [7.852941176470588, '07'], [38.5948275862069, '15'], [21.525, '20'], [10.08695652173913, '05']]


In [63]:
print("Top 5 Hours for Ask Posts Comments")

for average,hour in sorted_swap[0:5]:
    hour=dt.datetime.strptime(hour,"%H")
    hour=hour.strftime("%H:%M")
    print("{0}: {1:.2f} average comments per post".format(hour,average))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


In order to have higher chances to receive comments, one should create a post at: 3pm, 2am, 8pm, 4pm or 9pm.